File: scripts/extract-code-blocks.js
#!/usr/bin/env node
// Extract code blocks from article body fields into section.code arrays
const fs = require('fs');

const files = process.argv.slice(2);
const totalStats = {};

for (const filePath of files) {
  console.log(`\n=== Processing ${filePath} ===`);
  let raw = fs.readFileSync(filePath, 'utf-8');
  
  // Match code blocks: \`\`\`lang\n...\`\`\`
  // In raw file, each backtick is escaped as \ + backtick
  const codeBlockRe = /\\`\\`\\`(\w*)\n([\s\S]*?)\\`\\`\\`/g;
  
  const blocks = [];
  let m;
  while ((m = codeBlockRe.exec(raw)) !== null) {
    blocks.push({
      full: m[0],
      lang: m[1] || 'text',
      code: m[2],
      idx: m.index,
      end: m.index + m[0].length,
    });
  }
  
  console.log(`Found ${blocks.length} code blocks`);
  
  if (blocks.length === 0) continue;
  
  // Find section bodies: locate each `body: ` then find its content
  // A body is: body: `...content...`
  // The closing backtick is an unescaped ` followed by , or }
  
  // Find all body: ` positions
  const bodyRe = /body:\s*`/g;
  const bodyStarts = [];
  let bm;
  while ((bm = bodyRe.exec(raw)) !== null) {
    // The backtick is at bm.index + bm[0].length - 1
    bodyStarts.push(bm.index + bm[0].length - 1);
  }
  
  console.log(`Found ${bodyStarts.length} body fields`);
  
  // For each body, find its closing backtick
  // Scan from position+1, skip \`, stop at unescaped `
  const sections = [];
  for (const startIdx of bodyStarts) {
    let p = startIdx + 1;
    while (p < raw.length) {
      if (raw[p] === '\\' && raw[p + 1] === '`') {
        p += 2;
        continue;
      }
      if (raw[p] === '`' && raw[p + 1] === ',') {
        // Found closing backtick
        sections.push({
          start: startIdx + 1, // content starts after opening backtick
          end: p, // content ends at closing backtick
        });
        break;
      }
      p++;
    }
  }
  
  // Assign code blocks to sections
  for (const block of blocks) {
    for (let si = 0; si < sections.length; si++) {
      if (block.idx >= sections[si].start && block.end <= sections[si].end) {
        if (!sections[si].blocks) sections[si].blocks = [];
        sections[si].blocks.push(block);
        block.si = si;
        break;
      }
    }
  }
  
  // Print section summary
  for (let si = 0; si < sections.length; si++) {
    // Find section title
    const before = raw.substring(Math.max(0, sections[si].start - 300), sections[si].start);
    const tm = before.match(/title:\s*"([^"]*)"/);
    const title = tm ? tm[1] : '?';
    const bCount = (sections[si].blocks || []).length;
    console.log(`  Section "${title}": ${bCount} code blocks`);
  }
  
  // Process sections from end to start (preserve indices)
  let totalExtracted = 0;
  for (let si = sections.length - 1; si >= 0; si--) {
    const sec = sections[si];
    if (!sec.blocks || sec.blocks.length === 0) continue;
    
    // Reverse blocks for removal (highest index first)
    const orderedBlocks = [...sec.blocks].sort((a, b) => a.idx - b.idx);
    
    // Build code array
    const codeEntries = orderedBlocks.map(b => {
      const escapedCode = b.code.replace(/`/g, '\\`');
      return `    {\n      lang: "${b.lang}",\n      code: \`${escapedCode}\`,\n    }`;
    });
    
    const codeArrayStr = codeEntries.join(',\n');
    const newCodeField = `,\n      code: [\n${codeArrayStr}\n      ]`;
    
    // Remove code blocks from body (from end to start to preserve indices)
    let newBody = raw.substring(sec.start, sec.end);
    let offsetAdjustment = 0;
    
    for (let bi = orderedBlocks.length - 1; bi >= 0; bi--) {
      const block = orderedBlocks[bi];
      const relStart = block.idx - sec.start;
      const relEnd = block.end - sec.start;
      
      // Determine replacement text
      let replacement = '';
      if (block.lang === 'text' && (block.code.includes('┌') || block.code.includes('──') || block.code.includes('│'))) {
        replacement = '结构如下：';
      } else if (block.lang === 'text') {
        replacement = '流程如下：';
      } else if (block.lang === 'bash') {
        replacement = 'bash 代码如下：';
      } else if (block.lang === 'yaml') {
        replacement = 'yaml 代码如下：';
      } else if (block.lang === 'json') {
        replacement = 'json 代码如下：';
      } else {
        replacement = `${block.lang} 代码如下：`;
      }
      
      newBody = newBody.substring(0, relStart) + replacement + newBody.substring(relEnd);
    }
    
    // Find position to insert code field: after body closing backtick
    // The body closing backtick position in the full raw string
    const bodyCloseIdx = sec.end; // position of closing backtick
    // We need to insert after `, (backtick + comma)
    const insertAfter = bodyCloseIdx + 2; // skip ` and ,
    
    // Calculate the new insert position accounting for body changes
    const bodyDelta = newBody.length - (sec.end - sec.start);
    const actualInsertPos = insertAfter + bodyDelta;
    
    // Reconstruct the file
    // The new body replaces raw[sec.start:sec.end]
    raw = raw.substring(0, sec.start) + newBody + raw.substring(sec.end);
    
    // Now insert the code field
    // After adjustment, the position after bodyCloseIdx+2 shifts by bodyDelta
    const finalInsertPos = insertAfter + bodyDelta;
    raw = raw.substring(0, finalInsertPos) + newCodeField + raw.substring(finalInsertPos);
    
    totalExtracted += orderedBlocks.length;
  }
  
  fs.writeFileSync(filePath, raw);
  console.log(`  -> Extracted ${totalExtracted} code blocks`);
  totalStats[filePath] = totalExtracted;
}

console.log('\n=== Summary ===');
let grandTotal = 0;
for (const [f, c] of Object.entries(totalStats)) {
  console.log(`  ${f}: ${c}`);
  grandTotal += c;
}
console.log(`  Total: ${grandTotal}`);

File: scripts/extract-code-blocks.py
#!/usr/bin/env python3
"""
Extract \`\`\` code blocks from body fields in blog TS files.
The backticks in TS template literals are escaped as \`\`\` (literal backslash+backtick).
"""

import re
import sys
from pathlib import Path

# In the file, code fences are: \`\`\` (backslash+backtick x3) = hex 5c605c605c60
# Opening: line starting with \`\`\` optionally followed by lang
# Closing: line with \`\`\` possibly followed by backtick and comma

ESCAPE = r'\`\`\`'  # literal 6 chars: \`\`\`

def extract_blocks_from_body(body_text: str):
    """Extract code blocks from body text.
    Code block: \`\`\`lang\ncontent\n\`\`\`
    Returns list of (lang, code_content) and cleaned body text.
    """
    # Pattern: \`\`\`lang\n...\`\`\`
    pattern = re.escape(ESCAPE) + r'(\w*)\n(.*?)' + re.escape(ESCAPE)
    
    blocks = []
    def replacer(m):
        lang = m.group(1).strip() or 'text'
        code = m.group(2).strip('\n')
        blocks.append((lang, code))
        return ''
    
    new_body = re.sub(pattern, replacer, body_text, flags=re.DOTALL)
    # Clean up excessive blank lines
    new_body = re.sub(r'\n{3,}', '\n\n', new_body)
    new_body = new_body.strip('\n')
    
    return blocks, new_body


def process_file(filepath: str) -> int:
    content = Path(filepath).read_text(encoding='utf-8')
    
    # Strategy: process body fields one at a time
    # Find 'body: `' and match to closing '`'
    
    # Find all body: ` positions
    body_field_starts = []
    search_pos = 0
    while True:
        idx = content.find('body: `', search_pos)
        if idx == -1:
            break
        # The body content starts after 'body: `'
        body_start = idx + len('body: `')
        
        # Find the closing backtick
        # It's a backtick followed by comma, or backtick followed by newline
        closing = content.find('`', body_start)
        while closing != -1:
            # Check: is this the actual closing?
            # The closing backtick is the first unescaped backtick after content starts
            # But in TS template literals, backticks INSIDE are escaped with \
            # So an unescaped backtick is one NOT preceded by \
            if closing > 0 and content[closing-1] == '\\':
                # This backtick is escaped, skip it
                closing = content.find('`', closing + 1)
                continue
            break
        
        if closing == -1:
            break
        
        body_field_starts.append((idx, body_start, closing))
        search_pos = closing + 1
    
    total_extracted = 0
    
    # Process from end to start to preserve positions
    for field_idx, body_start, body_end in reversed(body_field_starts):
        body_text = content[body_start:body_end]
        
        # Extract code blocks
        blocks, new_body = extract_blocks_from_body(body_text)
        
        if not blocks:
            continue
        
        # Build the new code entries
        # Find indent
        line_start = content.rfind('\n', 0, field_idx)
        if line_start == -1:
            line_start = 0
        else:
            line_start += 1
        indent = content[line_start:field_idx]  # whitespace before "body:"
        
        code_indent = indent  # same as field indent
        entry_indent = code_indent + '  '
        detail_indent = entry_indent + '  '
        
        code_entries = []
        for lang, code_content in blocks:
            code_lines = code_content.split('\n')
            entry = f'{entry_indent}{{\n'
            entry += f'{entry_indent}lang: "{lang}",\n'
            if len(code_lines) == 1:
                entry += f'{entry_indent}code: `{code_lines[0]}`,\n'
            else:
                entry += f'{entry_indent}code: `\n'
                for cl in code_lines:
                    entry += f'{detail_indent}{cl}\n'
                entry += f'{entry_indent}`,\n'
            entry += f'{entry_indent}}}'
            code_entries.append(entry)
        
        # Check if existing code field follows
        after_body = content[body_end:]  # starts after closing backtick
        # Skip comma and whitespace
        after_body = after_body.lstrip(',\n ')
        
        has_code = bool(re.match(r'code:\s*\[', after_body))
        
        # Build replacement for the body field
        # body: `new_body`,
        if '\n' in new_body:
            new_body_field = f'body: `\n{new_body}\n{indent}`,'
        else:
            new_body_field = f'body: `{new_body}`,'
        
        if has_code:
            # Insert into existing code array
            # Find the code array boundaries
            code_array_start = body_end + len(content[body_end:]) - len(after_body)
            # Actually let's recalculate
            code_match = re.search(r'code:\s*\[', content[body_end:])
            if code_match:
                ca_start = body_end + code_match.start()
                # Find closing ],
                bracket = 0
                started = False
                ca_end = ca_start
                for ci in range(ca_start, len(content)):
                    if content[ci] == '[':
                        bracket += 1
                        started = True
                    elif content[ci] == ']':
                        bracket -= 1
                    if started and bracket == 0:
                        ca_end = ci + 1
                        break
                
                # Get existing code content
                existing_code = content[ca_start:ca_end]
                # Find where to insert (before the closing ])
                insert_pos = existing_code.rfind(']')
                existing_content = existing_code[:insert_pos]
                closing_bracket = existing_code[insert_pos:]
                
                # Clean up existing content - ensure trailing comma
                existing_content = existing_content.rstrip()
                if existing_content and not existing_content.endswith(','):
                    existing_content += ','
                
                new_code_array = existing_content + '\n' + '\n'.join(code_entries) + '\n' + closing_bracket
                full_replacement = new_body_field + content[body_end:ca_start] + new_code_array
                replace_start = field_idx
                replace_end = ca_end
            else:
                # Shouldn't happen since has_code was True
                new_code = f'\n{code_indent}code: [\n' + '\n'.join(code_entries) + f'\n{code_indent}],'
                full_replacement = new_body_field + new_code
                replace_start = field_idx
                replace_end = body_end + (1 if body_end < len(content) and content[body_end] == ',' else 0)
        else:
            new_code = f'\n{code_indent}code: [\n' + '\n'.join(code_entries) + f'\n{code_indent}],'
            full_replacement = new_body_field + new_code
            replace_start = field_idx
            # Include the comma after body if present
            replace_end = body_end
            if replace_end < len(content) and content[replace_end] == ',':
                replace_end += 1
        
        # Apply replacement
        content = content[:replace_start] + full_replacement + content[replace_end:]
        total_extracted += len(blocks)
    
    Path(filepath).write_text(content, encoding='utf-8')
    return total_extracted


if __name__ == '__main__':
    files = sys.argv[1:]
    for f in files:
        print(f"Processing {f}...")
        count = process_file(f)
        print(f"  Extracted {count} code block(s)")
